### PR TITLE
ColorPicker: TypeScript refactor

### PR DESCRIPTION
**Pull request by chad1008**
## What?
Refactor `ColorPicker` component's tests and stories to TypeScript

Part of https://github.com/WordPress/gutenberg/issues/35744

## Why?

The refactor to TypeScript has many benefits (auto-generated docs, static linting and error prevention, better IDE experience). See https://github.com/WordPress/gutenberg/issues/35744 for more details

For `ColorPicker`, the component itself is already written in TypeScript, tests and stories are all that need to be fully refactored.

## How?
Followed the steps in the [TypeScript migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md?rgh-link-date=2022-11-29T14%3A45%3A34Z#refactoring-a-component-to-typescript)

## Testing Instructions
- Ensure all unit and e2e tests still pass
- Test the component in the editor (e.g. on any color/background color selection) and Storybook to confirm there are no unexpected changes.

### CODE REVIEW

File: packages/components/src/color-picker/test/index.tsx [hunk 1]
```diff
@@ -81,8 +84,11 @@ describe( 'ColorPicker', () => {
 			);
 
 			const saturation = getSaturation( container );
+
+			expect( saturation ).toBeInTheDocument();
+
 			moveReactColorfulSlider(
-				saturation,
+				saturation!,
```
Comment (chad1008) [inline packages/components/src/color-picker/test/index.tsx]: The element here could be `null` if `querySelector` came up empty, so we should check for that.

To the best of my knowledge (though I'm happy if I'm wrong) TypeScript can't narrow based on test failures, so we need to narrow explicitly rule out `null` values now that they'll have been handled by the `toBeInTheDocument()` check.

This pattern is used twice more in subsequent tests, if a change is needed, I'll apply it to all three cases.

File: packages/components/src/color-picker/types.ts [hunk 2]
```diff
@@ -1,10 +1,117 @@
...
+	{
+		enableAlpha?: boolean;
+		color?: string;
+		onChange?: ( color: string ) => void;
+		defaultValue?: string;
+		copyFormat?: ColorType;
+	},
+	'div',
+	false
+>;
```
Comment (chad1008) [inline packages/components/src/color-picker/types.ts]: Previously, the argument you see here was typed in `component.tsx` and then immediately passed into `WordPressComponentProps` as `ColorPicker` was declared.

The resulting type was then accessed via `ComponentProps< typeof ColorPicker >` in `use-deprecated-props.ts`.

Now that this has been moved to the dedicated `types` file, I've consolidated that into a single declaration.

File: packages/components/src/color-picker/stories/index.tsx [hunk 3]
```diff
@@ -22,9 +25,13 @@ export default {
 		onChange: { action: 'onChange' },
 	},
 };
+export default meta;
 
-const Template = ( { onChange, ...props } ) => {
-	const [ color, setColor ] = useState();
+const Template: ComponentStory< typeof ColorPicker > = ( {
+	onChange,
```
Comment (chad1008) [inline packages/components/src/color-picker/stories/index.tsx:34]: TS errors here, because it can't distinguish between `LegacyProps` or `ColorPickerProps`. Need to investigate that further.

File: packages/components/src/color-picker/test/index.tsx [hunk 4]
```diff
@@ -138,8 +147,11 @@ describe( 'ColorPicker', () => {
 		);
```
Comment (chad1008) [inline packages/components/src/color-picker/test/index.tsx:149]: This test feels faulty to me, but I'd like an extra pair of 👀 .

TLDR: this test is passing `color` as
```
{ h: string; s: string; l: string; a?: string }
```

Looking at the current code/types, I don't see a case where the component actually accepts a value in that shape. It's typed to accept a `LegacyColor` which is much more elaborate and appears to be generated by utility functions.

When testing this manually by setting this object as the `color` value for the default story, it "works" but not well. The desired color isn't accurately represented and the component renders with an eight digit hex code... so it's trying to parse the alpha channel and not ignoring it. If you switch the component to HSL the saturation slider _does_ work. It's janky, but it works enough to fire `onChange`. So I see why the test has been passing, but I'm thinking we might want to rework this a bit.

File: packages/components/src/color-picker/test/index.tsx [hunk 5]
```diff
@@ -13,20 +13,19 @@ import { ColorPicker } from '..';
...
 }
 
+type PageXPageY = { pageX: number; pageY: number };
+
 // Fix to pass `pageX` and `pageY`
 // See https://github.com/testing-library/react-testing-library/issues/268
-class FakeMouseEvent extends window.MouseEvent {
-	constructor( type, values = {} ) {
+class FakeMouseEvent extends MouseEvent {
+	constructor( type: MouseEvent[ 'type' ], values: PageXPageY ) {
```
Comment (ciampo) [inline packages/components/src/color-picker/test/index.tsx]: Should we bring back the default value assignment to `{}`?

File: packages/components/src/color-picker/types.ts (hunk 2 above)
Comment (ciampo) [reply packages/components/src/color-picker/types.ts]: It looks like we're still using the `WordPressComponentProps` in `component.tsx`, we should probably remove it.

File: packages/components/src/color-picker/stories/index.tsx (hunk 3 above)
Comment (ciampo) [reply packages/components/src/color-picker/stories/index.tsx:34]: That is because `LegacyProps` doesn't list `onChange` at all.

<details>

<summary>One way to fix this would to be to more explicitly check that `onChange` is effectively defined on the `props` object</summary>

```diff
diff --git a/packages/components/src/color-picker/stories/index.tsx b/packages/components/src/color-picker/stories/index.tsx
index 29d8a43a42..0b2572c4cc 100644
--- a/packages/components/src/color-picker/stories/index.tsx
+++ b/packages/components/src/color-picker/stories/index.tsx
@@ -27,10 +27,7 @@ const meta: ComponentMeta< typeof ColorPicker > = {
 };
 export default meta;
 
-const Template: ComponentStory< typeof ColorPicker > = ( {
-	onChange,
-	...props
-} ) => {
+const Template: ComponentStory< typeof ColorPicker > = ( props ) => {
 	const [ color, setColor ] = useState< string | undefined >();
 
 	return (
@@ -38,7 +35,9 @@ const Template: ComponentStory< typeof ColorPicker > = ( {
 			{ ...props }
 			color={ color }
 			onChange={ ( ...changeArgs ) => {
-				onChange( ...changeArgs );
+				if ( 'onChange' in props ) {
+					props.onChange?.( ...changeArgs );
+				}
 				setColor( ...changeArgs );
 			} }
 		/>

```
</details>


<details>

<summary>Alternatively, we could define `onChange` on `LegacyProps` (using `never`), so that TypeScript "sees" that prop in both  `LegacyProps` and `ColorPickerProps`</summary>

```diff
diff --git a/packages/components/src/color-picker/stories/index.tsx b/packages/components/src/color-picker/stories/index.tsx
index 29d8a43a42..00ce542360 100644
--- a/packages/components/src/color-picker/stories/index.tsx
+++ b/packages/components/src/color-picker/stories/index.tsx
@@ -38,7 +38,7 @@ const Template: ComponentStory< typeof ColorPicker > = ( {
 			{ ...props }
 			color={ color }
 			onChange={ ( ...changeArgs ) => {
-				onChange( ...changeArgs );
+				onChange?.( ...changeArgs );
 				setColor( ...changeArgs );
 			} }
 		/>
diff --git a/packages/components/src/color-picker/types.ts b/packages/components/src/color-picker/types.ts
index 4cadb09bd5..6128d10548 100644
--- a/packages/components/src/color-picker/types.ts
+++ b/packages/components/src/color-picker/types.ts
@@ -114,4 +114,5 @@ export interface LegacyProps {
 	 * @deprecated
 	 */
 	disableAlpha: boolean;
+	onChange?: never;
 }

```
</details>

File: packages/components/src/color-picker/test/index.tsx (hunk 1 above)
Comment (ciampo) [reply packages/components/src/color-picker/test/index.tsx]: I wonder if adding an explicit check + throwing error may work (ie. the test would need to fail if the element is `null`)

```diff
diff --git a/packages/components/src/color-picker/test/index.tsx b/packages/components/src/color-picker/test/index.tsx
index 2820710268..80c287081e 100644
--- a/packages/components/src/color-picker/test/index.tsx
+++ b/packages/components/src/color-picker/test/index.tsx
@@ -85,10 +85,14 @@ describe( 'ColorPicker', () => {
 
 			const saturation = getSaturation( container );
 
+			if ( saturation === null ) {
+				throw new Error( 'The saturation slider could not be found' );
+			}
+
 			expect( saturation ).toBeInTheDocument();
 
 			moveReactColorfulSlider(
-				saturation!,
+				saturation,
 				{ pageX: 0, pageY: 0 },
 				{ pageX: 10, pageY: 10 }
 			);

```

File: packages/components/src/color-picker/test/index.tsx (hunk 4 above)
Comment (ciampo) [reply packages/components/src/color-picker/test/index.tsx:149]: Looking at the current scenario, the props that are being passed (`onChange`, `enableAlpha`) are for the new version of the component (ie. not the legacy one).

In that sense, it seems like `color` should be a string, and not an object?

We should dig a bit more into the history of this test. Was is written for the legacy color picker, and then just kept when migrating to the new one to make sure that we didn't introduce breaking changes? Or was the test added when introducing the newer version of the component?

In any case, we should try to assess if the test is actually testing a real, possible scenario (which means that our types may need updating), or if the test should be updated instead to reflect our types.

File: packages/components/src/color-picker/test/index.tsx (hunk 1 above)
Comment (chad1008) [reply packages/components/src/color-picker/test/index.tsx]: Smart, seems to work great. Thanks! f33002c694

File: packages/components/src/color-picker/types.ts (hunk 2 above)
Comment (chad1008) [reply packages/components/src/color-picker/types.ts]: Well spotted, thank you! 718a8b4713

File: packages/components/src/color-picker/stories/index.tsx (hunk 3 above)
Comment (chad1008) [reply packages/components/src/color-picker/stories/index.tsx:34]: Perfect. I went with the second option, it felt like cleaner and easier to parse. 787a52acba

File: packages/components/src/color-picker/test/index.tsx (hunk 5 above)
Comment (chad1008) [reply packages/components/src/color-picker/test/index.tsx]: Good point, thank you.

I needed to make some small changes alongside this to keep TS happy: 87ca35d8eb

Comment (chad1008) [reply packages/components/src/color-picker/test/index.tsx]: Actually, circling back to this, I think my previous fix failed to account for the possibility of `null` or `undefined` values, so I tweaked it a bit: e40cfc4867

File: packages/components/src/color-picker/test/index.tsx (hunk 4 above)
Comment (chad1008) [reply packages/components/src/color-picker/test/index.tsx:149]: Thanks for the input! The history is interesting. It looks like this test was added in #35220 when the component was replaced with the new version. At that point in time, the component's `color` prop was still an HSL/HSLA object using a type drawn from `tinycolor2`, which matches the object currently being passed in this test. So the test was valid and informative when it was written.

Then in #35562 we removed `tinycolor2` and in that PR included a breaking change notice in the changelog, stating the component should now accept a string.

So based on all of that, I think this test is trying to validate something we no longer actually expect the component to do, and even though it's technically passing (aside from the TS error) it probably makes sense to remove it. Thoughts?

File: packages/components/src/color-picker/test/index.tsx (hunk 5 above)
Comment (ciampo) [reply packages/components/src/color-picker/test/index.tsx]: I took another look at this, and I actually think that your initial approach of removing the default `{}` value may be actually be cleaner — I would just tweak it so that `values` can be optional (which is probably the reason why we had that default value of `{}`).

We can greatly simplify the code by using optional chaining:

```diff
diff --git a/packages/components/src/color-picker/test/index.tsx b/packages/components/src/color-picker/test/index.tsx
index addda8c6b4..9ef79cde2e 100644
--- a/packages/components/src/color-picker/test/index.tsx
+++ b/packages/components/src/color-picker/test/index.tsx
@@ -25,16 +25,12 @@ type PageXPageY = { pageX: number; pageY: number };
 // Fix to pass `pageX` and `pageY`
 // See https://github.com/testing-library/react-testing-library/issues/268
 class FakeMouseEvent extends MouseEvent {
-	constructor( type: MouseEvent[ 'type' ], values: PageXPageY | {} = {} ) {
+	constructor( type: MouseEvent[ 'type' ], values?: PageXPageY ) {
 		super( type, { buttons: 1, bubbles: true, ...values } );
 
-		const parseXYValue = ( value: number ) => {
-			return value ?? 0;
-		};
-
 		Object.assign( this, {
-			pageX: 'pageX' in values ? parseXYValue( values.pageX ) : 0,
-			pageY: 'pageY' in values ? parseXYValue( values.pageY ) : 0,
+			pageX: values?.pageX ?? 0,
+			pageY: values?.pageY ?? 0,
 		} );
 	}
 }

```

File: packages/components/src/color-picker/test/index.tsx (hunk 4 above)
Comment (ciampo) [reply packages/components/src/color-picker/test/index.tsx:149]: Thank you for tracking back the history of this test.

I think that the best thing to do, for the sake of this PR specifically, is to keep the test and simply convert the format of the color value, ie.

```diff
diff --git a/packages/components/src/color-picker/test/index.tsx b/packages/components/src/color-picker/test/index.tsx
index addda8c6b4..2371d32d06 100644
--- a/packages/components/src/color-picker/test/index.tsx
+++ b/packages/components/src/color-picker/test/index.tsx
@@ -142,13 +142,7 @@ describe( 'ColorPicker', () => {
 
 	it( 'should fire onChange with the HSL value', async () => {
 		const onChange = jest.fn();
-		const color = {
-			h: 125,
-			s: 0.2,
-			l: 0.5,
-			// Add alpha to prove it's ignored.
-			a: 0.5,
-		};
+		const color = 'hsla(125, 20%, 50%, 0.5)';
 
 		const { container } = render(
 			<ColorPicker
```

While taking a look at this, though, I noticed that the unit test for the component are not particularly detailed — for example:

- they don't use `user-event`
- they don't check for the exact color values that are passed to the `onChange` function (they currently only check that it's a valid hex value)
- they don't check the value of the input fields
- they only interact with the component by simulating a mouse drag over the "canvas", and therefore they don't test all of the different input fields available:
  - The hex input
  - The HSL input/sliders
  - The RGB input/sliders
  - The hue slider
  - The Alpha slider

I feel like we should work on improving those tests in a follow-up, since this component is quite important to the editor.

File: packages/components/src/color-picker/stories/index.tsx [hunk 6]
```diff
@@ -1,14 +1,17 @@
...
  */
 import { useState } from '@wordpress/element';
-
 /**
  * Internal dependencies
  */
 import { ColorPicker } from '..';
 
-export default {
+const meta: ComponentMeta< typeof ColorPicker > = {
```
Comment (ciampo) [inline packages/components/src/color-picker/stories/index.tsx:16]: We should leverage TS for generating docs, but at the same time, we should hide those legacy props — something like this may work

```diff
diff --git a/packages/components/src/color-picker/stories/index.tsx b/packages/components/src/color-picker/stories/index.tsx
index 00ce542360..d9f8b0db72 100644
--- a/packages/components/src/color-picker/stories/index.tsx
+++ b/packages/components/src/color-picker/stories/index.tsx
@@ -20,10 +20,32 @@ const meta: ComponentMeta< typeof ColorPicker > = {
 			control: { type: 'select' },
 			options: [ 'rgb', 'hsl', 'hex', undefined ],
 		},
+		onChangeComplete: {
+			table: {
+				disable: true,
+			},
+		},
+		oldHue: {
+			table: {
+				disable: true,
+			},
+		},
+		disableAlpha: {
+			table: {
+				disable: true,
+			},
+		},
 		// We can't use a `on*` regex because this component will switch to
 		// legacy mode when an onChangeComplete prop is passed.
 		onChange: { action: 'onChange' },
 	},
+	parameters: {
+		controls: {
+			expanded: true,
+		},
+		docs: { source: { state: 'open' } },
+	},
 };
 export default meta;

```

File: packages/components/src/color-picker/stories/index.tsx
Comment (ciampo) [inline packages/components/src/color-picker/stories/index.tsx:1]: I also noticed that the props in the docs table in Storybook don't currently have a description, could we add them via JSDocs in the `types.ts` file?

File: packages/components/src/color-picker/test/index.tsx (hunk 4 above)
Comment (chad1008) [reply packages/components/src/color-picker/test/index.tsx:149]: I was thinking they looked a bit sparse as well. I actually started looking at how to simulate the mouse movement under `user-event` before stepping back to focus on the TypeScript bits.

Implemented the above change for now: 7458897813

File: packages/components/src/color-picker/stories/index.tsx (hunk 6 above)
Comment (chad1008) [reply packages/components/src/color-picker/stories/index.tsx:16]: ✅  feba5ddcb2

File: packages/components/src/color-picker/stories/index.tsx
Comment (chad1008) [reply packages/components/src/color-picker/stories/index.tsx:1]: ✅ 7340455d0e

File: packages/components/src/color-picker/test/index.tsx (hunk 5 above)
Comment (chad1008) [reply packages/components/src/color-picker/test/index.tsx]: Yes, that's much better. b0e9423679

**Comment by chad1008** [comment]
Addressed latest feedback and updated PR to no longer be a draft

File: packages/components/src/color-picker/stories/index.tsx (hunk 6 above)
Comment (ciampo) [reply packages/components/src/color-picker/stories/index.tsx:16]: Actually, it still looks like we could do better — what if, instead of using the legacy adapter, we used the new picker in Storybook? That way, we'd get automatically the docs for just the new props, and we would be able to delete the "hacky" configs that I suggested above.

<details>

<summary>Something like this maybe?</summary>

```diff
diff --git a/packages/components/src/color-picker/component.tsx b/packages/components/src/color-picker/component.tsx
index 379b8557d0..852e0756b3 100644
--- a/packages/components/src/color-picker/component.tsx
+++ b/packages/components/src/color-picker/component.tsx
@@ -38,7 +38,7 @@ const options = [
 	{ label: 'Hex', value: 'hex' as const },
 ];
 
-const ColorPicker = (
+const UnconnectedColorPicker = (
 	props: ColorPickerProps,
 	forwardedRef: ForwardedRef< any >
 ) => {
@@ -112,6 +112,9 @@ const ColorPicker = (
 	);
 };
 
-const ConnectedColorPicker = contextConnect( ColorPicker, 'ColorPicker' );
+export const ColorPicker = contextConnect(
+	UnconnectedColorPicker,
+	'ColorPicker'
+);
 
-export default ConnectedColorPicker;
+export default ColorPicker;
diff --git a/packages/components/src/color-picker/stories/index.tsx b/packages/components/src/color-picker/stories/index.tsx
index bb4adb5410..419c719571 100644
--- a/packages/components/src/color-picker/stories/index.tsx
+++ b/packages/components/src/color-picker/stories/index.tsx
@@ -2,44 +2,26 @@
  * External dependencies
  */
 import type { ComponentMeta, ComponentStory } from '@storybook/react';
+
 /**
  * WordPress dependencies
  */
 import { useState } from '@wordpress/element';
+
 /**
  * Internal dependencies
  */
-import { ColorPicker } from '..';
+import { ColorPicker } from '../component';
 
 const meta: ComponentMeta< typeof ColorPicker > = {
 	component: ColorPicker,
 	title: 'Components/ColorPicker',
 	argTypes: {
+		as: { control: { type: null } },
 		color: { control: { type: null } },
-		copyFormat: {
-			control: { type: 'select' },
-			options: [ 'rgb', 'hsl', 'hex', undefined ],
-		},
-		onChangeComplete: {
-			table: {
-				disable: true,
-			},
-		},
-		oldHue: {
-			table: {
-				disable: true,
-			},
-		},
-		disableAlpha: {
-			table: {
-				disable: true,
-			},
-		},
-		// We can't use a `on*` regex because this component will switch to
-		// legacy mode when an onChangeComplete prop is passed.
-		onChange: { action: 'onChange' },
 	},
 	parameters: {
+		actions: { argTypesRegex: '^on.*' },
 		controls: {
 			expanded: true,
 		},

```
</details>

File: packages/components/src/color-picker/README.md [hunk 7]
```diff
@@ -31,13 +31,13 @@ The current color value to display in the picker. Must be a hex or hex8 string.
 
 ### `onChange`: `(hex8Color: string) => void`
 
-Fired when the color changes. Always passes a hex8 color string.
+Fired when the color changes. Always passes a hex or hex8 color string.
```
Comment (chad1008) [inline packages/components/src/color-picker/README.md:34]: `onChange` previously always passed a hex8 string, but since #35562, `Colord.toHex()` will return a hex or a hex8 (hex8 only if the color's alpha is < 1)

File: packages/components/src/color-picker/stories/index.tsx (hunk 6 above)
Comment (ciampo) [reply packages/components/src/color-picker/stories/index.tsx:16]: I pushed the suggested changes while Chad is AFK

**Review by ciampo — APPROVED** [review]
🚀 

- No runtime changes
- Types moved to the same `types.ts` file
- Better Storybook examples and auto-generated docs
- Updated some prop descriptions and a unit tests that was still relying on the old color format

File: packages/components/src/color-picker/README.md (hunk 7 above)
Comment (ciampo) [reply packages/components/src/color-picker/README.md:34]: Nice catch, thank you!
